### PR TITLE
Add residual collapsing of Conv2D

### PR DESCRIPTION
**Pull request by eladc-git**
Support collapsing add residual to Conv2D layer.
Keras+Pytorch.

### CODE REVIEW

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/residual_collapsing.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    second_node = NodeOperationMatcher(torch.add)
+    return first_node, second_node
+
+
+def residual_collapsing_fn(first_node: BaseNode,
+                           kernel_str: str) -> np.ndarray:
+    """
+    # Collapsing residual addition into convolution
+    # Inspired by https://arxiv.org/pdf/2103.09404.pdf - Algorithm2
+    Args:
```
Comment (reuvenperetz) [inline model_compression_toolkit/pytorch/graph_substitutions/substitutions/residual_collapsing.py:42]: Please fix comments

File: tests/keras_tests/feature_networks_tests/feature_networks/residual_collapsing_test.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+# Copyright 2021 Sony Semiconductors Israel, Inc. All rights reserved.
+#
```
Comment (reuvenperetz) [inline tests/keras_tests/feature_networks_tests/feature_networks/residual_collapsing_test.py:2]: Fix year

File: tests/pytorch_tests/model_tests/feature_models/residual_collapsing_test.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+from model_compression_toolkit.tpc_models.pytorch_tp_models.pytorch_default import generate_pytorch_tpc
+from tests.common_tests.helpers.tensors_compare import cosine_similarity
+
+tp = mct.target_platform
+
+class BaseResidualCollapsingTest(BasePytorchFeatureNetworkTest, ABC):
+
+    def __init__(self, unit_test):
+        super().__init__(unit_test=unit_test, input_shape=(16,32,32))
+
```
Comment (reuvenperetz) [inline tests/pytorch_tests/model_tests/feature_models/residual_collapsing_test.py:32]: Fix ABC and input shape, please

File: tests/pytorch_tests/model_tests/feature_models/residual_collapsing_test.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    def compare(self, quantized_model, float_model, input_x=None, quantization_info=None):
+        in_torch_tensor = to_torch_tensor(input_x[0])
+        y = float_model(in_torch_tensor)
+        y_hat = quantized_model(in_torch_tensor)
+        self.unit_test.assertTrue(y.shape == y_hat.shape, msg=f'out shape is not as expected!')
+        for layer in quantized_model.children():
+            self.unit_test.assertFalse(type(layer) == torch.add, msg=f'fail: add residual is still in the model')
+        cs = cosine_similarity(torch_tensor_to_numpy(y), torch_tensor_to_numpy(y_hat))
+        self.unit_test.assertTrue(np.isclose(cs, 1), msg=f'fail: cosine similarity check:{cs}')
+
```
Comment (reuvenperetz) [inline tests/pytorch_tests/model_tests/feature_models/residual_collapsing_test.py:53]: Same remark as in Keras

**Review by reuvenperetz — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/common/post_training_quantization.py [hunk 5]
```diff
@@ -209,8 +209,9 @@ def get_finalized_graph(initial_graph: Graph,
     # Graph substitution (pre statistics collection)
     ######################################
     transformed_graph = substitute(graph, fw_impl.get_substitutions_pre_statistics_collection(quant_config))
-    transformed_graph = linear_collapsing_substitute(transformed_graph, fw_impl.get_linear_collapsing_substitution(), quant_config.block_collapsing)
-
+    if quant_config.block_collapsing:
+        transformed_graph = linear_collapsing_substitute(transformed_graph, fw_impl.get_linear_collapsing_substitution())
+        transformed_graph = substitute(transformed_graph, fw_impl.get_residual_collapsing_substitution())
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/common/post_training_quantization.py:215]: Why? it's over-parametrization

File: model_compression_toolkit/keras/graph_substitutions/substitutions/residual_collapsing.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    first_node = first_node & activation_linear
+    return first_node, second_node
+
+
+def residual_collapsing_fn(first_node: BaseNode,
+                           kernel_str: str) -> np.ndarray:
+    """
+    # Collapsing residual addition into
+    # Inspired by https://arxiv.org/pdf/2103.09404.pdf - Algorithm2
+    Args:
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/keras/graph_substitutions/substitutions/residual_collapsing.py:44]: Done

File: tests/keras_tests/feature_networks_tests/feature_networks/residual_collapsing_test.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+                                      False, False, True)
+
+    def compare(self, quantized_model, float_model, input_x=None, quantization_info=None):
+        y = float_model.predict(input_x)
+        y_hat = quantized_model.predict(input_x)
+        self.unit_test.assertTrue(y.shape == y_hat.shape, msg=f'fail: out shape is not as expected!')
+        for layer in quantized_model.layers:
+            self.unit_test.assertFalse(type(layer) == layers.Add, msg=f'fail: add residual is still in the model')
+        cs = cosine_similarity(y, y_hat)
+        self.unit_test.assertTrue(np.isclose(cs, 1), msg=f'fail: cosine similarity check:{cs}')
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply tests/keras_tests/feature_networks_tests/feature_networks/residual_collapsing_test.py:53]: this is done by other metric: np.isclose(cs, 1)
I add second assert to check sanity - no "add layer" need to be here. This is **consistent** with previous PR which you request another assert for sanity and not only for outout.

File: tests/keras_tests/feature_networks_tests/feature_networks/residual_collapsing_test.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+keras = tf.keras
+layers = keras.layers
+tp = mct.target_platform
+
+
+class BaseResidualCollapsingTest(BaseKerasFeatureNetworkTest, ABC):
+
+    def __init__(self, unit_test):
+        super(BaseResidualCollapsingTest, self).__init__(unit_test=unit_test, input_shape=(32,32,16))
+
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply tests/keras_tests/feature_networks_tests/feature_networks/residual_collapsing_test.py:34]: Done

File: tests/keras_tests/feature_networks_tests/feature_networks/residual_collapsing_test.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+import numpy as np
+from tests.common_tests.helpers.tensors_compare import cosine_similarity
+
+keras = tf.keras
+layers = keras.layers
+tp = mct.target_platform
+
+
+class BaseResidualCollapsingTest(BaseKerasFeatureNetworkTest, ABC):
+
```
[the remark this answers is not in the document]
Comment (eladc-git) [reply tests/keras_tests/feature_networks_tests/feature_networks/residual_collapsing_test.py:31]: Done

File: model_compression_toolkit/pytorch/graph_substitutions/substitutions/residual_collapsing.py (hunk 1 above)
Comment (eladc-git) [reply model_compression_toolkit/pytorch/graph_substitutions/substitutions/residual_collapsing.py:42]: Done

File: tests/keras_tests/feature_networks_tests/feature_networks/residual_collapsing_test.py (hunk 2 above)
Comment (eladc-git) [reply tests/keras_tests/feature_networks_tests/feature_networks/residual_collapsing_test.py:2]: Done

File: tests/pytorch_tests/model_tests/feature_models/residual_collapsing_test.py (hunk 3 above)
Comment (eladc-git) [reply tests/pytorch_tests/model_tests/feature_models/residual_collapsing_test.py:32]: Done

File: tests/pytorch_tests/model_tests/feature_models/residual_collapsing_test.py (hunk 4 above)
Comment (eladc-git) [reply tests/pytorch_tests/model_tests/feature_models/residual_collapsing_test.py:53]: same response

File: model_compression_toolkit/common/post_training_quantization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (reuvenperetz) [reply model_compression_toolkit/common/post_training_quantization.py:215]: In general: it's bad practice to give multiple responsibilities to a single thing (class, function, variable)
https://en.wikipedia.org/wiki/Single-responsibility_principle

File: model_compression_toolkit/keras/graph_substitutions/substitutions/residual_collapsing.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+def residual_collapsing_node_matchers() -> Tuple[NodeOperationMatcher, NodeOperationMatcher]:
+    """
+    Function generates matchers for matching:
+    (Conv2D, Add)[activation=linear] -> Conv2D.
+    Returns:
+        Matcher for Conv2D node and Add node
+    """
+    first_node = NodeOperationMatcher(Conv2D)
+    second_node = NodeOperationMatcher(Add)
+    activation_linear = NodeFrameworkAttrMatcher(ACTIVATION, LINEAR)
```
Comment (reuvenperetz) [inline model_compression_toolkit/keras/graph_substitutions/substitutions/residual_collapsing.py:34]: Please regard the case of tf.add

Comment (eladc-git) [reply model_compression_toolkit/keras/graph_substitutions/substitutions/residual_collapsing.py:34]: Done

File: model_compression_toolkit/common/post_training_quantization.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (eladc-git) [reply model_compression_toolkit/common/post_training_quantization.py:216]: Done
